### PR TITLE
Add lint rule to check that `Debug.assert` calls do not eagerly interpolate strings

### PR DESCRIPTION
**Pull request by andy-ms**
The second argument is usually never used (unless an assertion fails), so we shouldn't spend time eagerly computing its value.

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -1285,12 +1285,12 @@ namespace ts {
     export function createFileDiagnostic(file: SourceFile, start: number, length: number, message: DiagnosticMessage): Diagnostic {
         const end = start + length;
 
-        Debug.assert(start >= 0, "start must be non-negative, is " + start);
-        Debug.assert(length >= 0, "length must be non-negative, is " + length);
+        Debug.assert(start >= 0, "start must be non-negative", () => `is ${start}`);
```
Comment (rbuckton) [inline src/compiler/core.ts]: Adding callbacks is arguably more expensive, as you have to create an environment record allocation each time.

It might make more sense to add something like `Debug.assertNonNegative(start, "start")` or similar.

File: src/compiler/transformers/generators.ts [hunk 2]
```diff
@@ -2448,7 +2448,7 @@ namespace ts {
          * @param location An optional source map location for the statement.
          */
         function createInlineBreak(label: Label, location?: TextRange): ReturnStatement {
-            Debug.assert(label > 0, `Invalid label: ${label}`);
+            Debug.assert(label > 0, "Invalid label", () => label.toString());
```
Comment (rbuckton) [inline src/compiler/transformers/generators.ts]: The callback is probably more expensive.

File: src/compiler/core.ts [hunk 3]
```diff
@@ -1285,12 +1285,12 @@ namespace ts {
     export function createFileDiagnostic(file: SourceFile, start: number, length: number, message: DiagnosticMessage): Diagnostic {
         const end = start + length;
 
-        Debug.assert(start >= 0, "start must be non-negative, is " + start);
-        Debug.assert(length >= 0, "length must be non-negative, is " + length);
+        Debug.assertLessThanOrEqual(0, start);
```
Comment (rbuckton) [inline src/compiler/core.ts]: assertGreaterThanOrEqual?

File: src/compiler/core.ts [hunk 4]
```diff
@@ -2334,15 +2334,34 @@ namespace ts {
             return currentAssertionLevel >= level;
         }
 
-        export function assert(expression: boolean, message?: string, verboseDebugInfo?: () => string, stackCrawlMark?: Function): void {
+        export function assert(expression: boolean, message?: string, verboseDebugInfo?: string | (() => string), stackCrawlMark?: Function): void {
```
Comment (rbuckton) [inline src/compiler/core.ts]: Why this change?

File: src/server/project.ts [hunk 5]
```diff
@@ -362,7 +362,7 @@ namespace ts.server {
             return map(this.program.getSourceFiles(), sourceFile => {
                 const scriptInfo = this.projectService.getScriptInfoForPath(sourceFile.path);
                 if (!scriptInfo) {
-                    Debug.assert(false, `scriptInfo for a file '${sourceFile.fileName}' is missing.`);
+                    Debug.assert(false, "scriptInfo for a file is missing:", sourceFile.fileName);
```
Comment (rbuckton) [inline src/server/project.ts]: No need to change this, other than from an `assert(false, ` to `fail` as the condition was already checked.

File: src/compiler/core.ts (hunk 4 above)
Comment (andy-ms) [inline src/compiler/core.ts]: This avoids needing a closure for the second argument (if it is not an interpolation).
